compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -2599,7 +2599,32 @@ object Types {
     final override def toString = s"$prefixString($paramNames, $paramInfos, $resType)"
   }
 
-  trait HKLambda extends LambdaType
+  abstract class HKLambda extends CachedProxyType with LambdaType {
+    final override def underlying(implicit ctx: Context) = resType
+
+    final override def computeHash = doHash(paramNames, resType, paramInfos)
+
+    final override def equals(that: Any) = that match {
+      case that: HKLambda =>
+        paramNames == that.paramNames &&
+        paramInfos == that.paramInfos &&
+        resType == that.resType &&
+        companion.eq(that.companion)
+      case _ =>
+        false
+    }
+
+    final override def eql(that: Type) = that match {
+      case that: HKLambda =>
+        paramNames.equals(that.paramNames) &&
+        paramInfos.equals(that.paramInfos) &&
+        resType.equals(that.resType) &&
```

Comment:
That was already the case before #3239, but why is this equals and not eq?

Reply:
I am not even sure, `equals` is needed. But the idea is that, since many types get a compiler generated `equals` since they are case classes, the rest of the types would conform to that.